File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/controllers/BatchHelperResource.java
package uk.gov.justice.digital.hmpps.riskprofiler.controllers;

import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiParam;
import io.swagger.annotations.ApiResponse;
import io.swagger.annotations.ApiResponses;
import io.swagger.annotations.Authorization;
import org.springframework.http.MediaType;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import uk.gov.justice.digital.hmpps.riskprofiler.events.QueueAdminService;
import uk.gov.justice.digital.hmpps.riskprofiler.model.ErrorResponse;
import uk.gov.justice.digital.hmpps.riskprofiler.model.EscapeProfile;
import uk.gov.justice.digital.hmpps.riskprofiler.model.SocProfile;
import uk.gov.justice.digital.hmpps.riskprofiler.schedule.PollPrisonersScheduler;
import uk.gov.justice.digital.hmpps.riskprofiler.services.PrisonService;

import javax.validation.constraints.NotNull;

@Api(tags = {"batch"},
        authorizations = {@Authorization("RISK_PROFILER")},
        description = "Provides ability to configure and run batch")

@RestController
@RequestMapping(
        value = "batch-helper",
        produces = MediaType.APPLICATION_JSON_VALUE)
public class BatchHelperResource {

    private final PollPrisonersScheduler pollPrisonersScheduler;
    private final PrisonService prisonService;
    private final QueueAdminService queueAdminService;

    public BatchHelperResource(final PollPrisonersScheduler pollPrisonersScheduler, final PrisonService prisonService, final QueueAdminService queueAdminService) {
        this.pollPrisonersScheduler = pollPrisonersScheduler;
        this.prisonService = prisonService;
        this.queueAdminService = queueAdminService;
    }

    @ApiOperation(value = "Start a batch job run", authorizations = {@Authorization("RISK_PROFILER")})
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = "OK"),
            @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
    @PreAuthorize("hasRole('RISK_PROFILER')")
    @PostMapping(path = "/startPollPrisoners")
    public void startPollPrisoners() {
        pollPrisonersScheduler.pollPrisoners();
        // could use camel if we want asych kickoff:
        // producerTemplate.send("direct:poll-prisoners", exchange -> {});
    }

    @ApiOperation(value = "Add prison to the uk.gov.justice.digital.hmpps.config", authorizations = {@Authorization("RISK_PROFILER")})
    @ApiResponses(value = {
            @ApiResponse(code = 200, message = "OK"),
            @ApiResponse(code = 400, message = "Invalid prison id", response = ErrorResponse.class),
            @ApiResponse(code = 500, message = "Unrecoverable error occurred whilst processing request.", response = ErrorResponse.class)})
    @PreAuthorize("hasRole('RISK_PROFILER')")
    @PostMapping(path = "/prison/{prisonId}")
    public void addPrison(@ApiParam(name = "prisonId", value = "Agency id of the prison", example = "LEI", required = true) @NotNull @PathVariable("prisonId") final String prisonId) {
        prisonService.addPrison(prisonId);
    }

    @ApiOperation(value = "Trigger the transfer of any DLQ messages to the event queue", authorizations = {@Authorization("RISK_PROFILER")})
    @PreAuthorize("hasRole('RISK_PROFILER')")
    @PostMapping(path = "/transferEventMessages")
    public void transferEventMessages() {
        queueAdminService.transferEventMessages();
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/controllers/BatchHelperResource.kt
package uk.gov.justice.digital.hmpps.riskprofiler.controllers

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.MediaType
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import uk.gov.justice.digital.hmpps.riskprofiler.events.QueueAdminService
import uk.gov.justice.digital.hmpps.riskprofiler.model.ErrorResponse
import uk.gov.justice.digital.hmpps.riskprofiler.schedule.PollPrisonersScheduler
import uk.gov.justice.digital.hmpps.riskprofiler.services.PrisonService
import javax.validation.constraints.NotNull

@Tag(
  name = "batch",
  description = "Provides ability to configure and run batches"
)
@ApiResponses(
  value = [
    ApiResponse(responseCode = "200", description = "OK"),
    ApiResponse(
      responseCode = "400",
      description = "Invalid prison id.",
      content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
    ),
    ApiResponse(
      responseCode = "500",
      description = "Unrecoverable error occurred whilst processing request.",
      content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))]
    ),
  ]
)

@RestController
@RequestMapping(value = ["batch-helper"], produces = [MediaType.APPLICATION_JSON_VALUE])
class BatchHelperResource(
  private val pollPrisonersScheduler: PollPrisonersScheduler,
  private val prisonService: PrisonService,
  private val queueAdminService: QueueAdminService
) {
  @Operation(summary = "Start a batch job run")
  @PreAuthorize("hasRole('RISK_PROFILER')")
  @PostMapping(path = ["/startPollPrisoners"])
  fun startPollPrisoners() {
    pollPrisonersScheduler.pollPrisoners()
    // could use camel if we want async kickoff:
    // producerTemplate.send("direct:poll-prisoners", exchange -> {});
  }

  @Operation(
    summary = "Add prison to the config",
    description = "The overnight polling batch will then include this prison"
  )
  @PreAuthorize("hasRole('RISK_PROFILER')")
  @PostMapping(path = ["/prison/{prisonId}"])
  fun addPrison(
    @Parameter(
      name = "prisonId",
      description = "Agency id of the prison",
      example = "LEI",
      required = true
    ) @PathVariable("prisonId") prisonId: @NotNull String
  ) =
    prisonService.addPrison(prisonId)

  @Operation(summary = "Trigger the transfer of any DLQ messages back to the event queue to be retried")
  @PreAuthorize("hasRole('RISK_PROFILER')")
  @PostMapping(path = ["/transferEventMessages"])
  fun transferEventMessages() =
    queueAdminService.transferEventMessages()
}
